proposals/p3833.md

```diff
@@ -0,0 +1,157 @@
+# SemIR fidelity when representing rewrite semantics
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/3833)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+    -   [Work to incorporate needed simplifications directly into the design](#work-to-incorporate-needed-simplifications-directly-into-the-design)
+    -   [Keep a full fidelity mode for any other optimizations](#keep-a-full-fidelity-mode-for-any-other-optimizations)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Strictly adhere to a full-fidelity model](#strictly-adhere-to-a-full-fidelity-model)
+    -   [Directly implement the optimized model](#directly-implement-the-optimized-model)
+
+<!-- tocstop -->
+
+## Abstract
+
+The toolchain's [Semantic IR][semir] should start off modeling the full,
+complex, and rich library-based and generic extension point semantics of Carbon
+without eliding any layers or rewrites for compile time efficiency. We shouldn't
+front-load elision or optimization when implementing the designs.
```

Comment:
Is the emphasis here "for compile time efficiency"? e.g., we've elided layers in order to reduce implementation cost -- is that okay? Or what does it mean to "start off" modeling everything -- does it really mean, it won't be considered complete until it models everything?

Note too, we do also optimize a little as we go along. I think there's a balance here where we probably don't want to spend too much time, but we should probably front-load a little.

Reply:
For a proposal, maybe the focus should be more on the goal of what the toolchain should be doing at milestones, rather than the implementation and how it evolves to get there?

Reply:
Not sure I understand this enough to address it based on the review alone -- might be worth having a VC discussion around it.

First off -- I'm not sure a proposal is the right way to document this, was an idea from @zygoloid and I figured we could try it. If a different model works, happy with that.

Secondly, definitely don't want to preclude whatever is needed when making incremental progress towards an implementation. I've added an example as suggested by Josh below and clarified this point in the example but happy to add other clarifications around that.

Beyond that, I think the example may also answer some of this. If it does, happy to adjust the rest of the wording to be more clear?

Reply:
A discussion might help. It sounds like this is deliberately making an implementation choice; I would like to understand the review process for future toolchain implementation choices, and it would probably be good to discuss checking in the toolchain architecture doc since this proposal would be the first thing deciding how the toolchain (and SemIR specifically) works.